rfd/0095-teleport-key-agent-extension.md

```diff
@@ -0,0 +1,175 @@
+
+### Teleport Key Agent Changes
+
+We will add new [SSH Agent extensions](https://datatracker.ietf.org/doc/html/draft-miller-ssh-agent#section-4.7) to forward `~/.tsh` profile information, certificates, and additional functions (per-session MFA). Basic SSH Agent functionality will not be impacted by adding these extensions, so existing integrations will remain unchanged.
+
+To support these extensions, `tsh` will need to be updated to extend the current Teleport Key Agent and add a new Teleport Key Agent client. The Teleport Key Agent client can be used by `tsh` calls on a remote host to send SSH Agent requests to the Teleport Key Agent running on the local host.
+
+#### SSH Agent Extensions
+
+**Extension Structure:**
+
+Extension requests consist of:
+
+        byte            SSH_AGENTC_EXTENSION
+        string          extension type (type@domain)
+        byte[]          extension contents
+
+An Extension response can be any custom message, but failures should result in `SSH_AGENT_EXTENSION_FAILURE`. Unsupported extensions should result in an `SSH_AGENT_FAILURE` response to differentiate from actual extension failures.
+
+**List Profiles Extension:**
+
+This extension requests `tsh` profile information, including the current profile and profile yaml files.
+
+        byte            SSH_AGENTC_EXTENSION
+        string          list-profiles@goteleport.com  
+
+The returned profile information can be used by `tsh` to initiate new Teleport clients.
+
+        byte            SSH_AGENT_SUCCESS
+        string          current profile name
+        byte[]          profiles blob (json)
+
+Where profiles blob is a json encoded `[]api/profile.Profile`.
+
+Note: The underlying `api/profile.Profile` struct is subject to change, so there is a possibility of backwards compatibility concerns. However, this backwards compatibility should be handled already since the same issue exists for `~/.tsh` profile yaml files.
+
+**List Keys Extension:**
+
+This extension requests a list of keys matching the given key index. Partial key indexes can be provided to return partial matches.
+
+        byte            SSH_AGENTC_EXTENSION
+        string          list-keys@goteleport.com
+        keyIndex[]      key index
+
+Where a keyIndex consists of:
+
+        string          proxy host       
+        string          cluster name
+        string          username
+
+The returned keys can be used in concert with the signers available through the SSH agent to perform TLS and SSH handshakes.
+
+        byte            SSH_AGENT_SUCCESS
+        key[]           keys
+        byte[]          known hosts
+
+Where a key consists of:
+
+        keyIndex        key index
+        byte[]          SSH certificate
```

Comment:
Will it be possible to list all end user's profiles, or will be there any restriction on forwarded certificates? (Could the end user restrict forwarded certificates from the /.tsh folder to required profiles only?) What happens in case of several hops, on every hop the output of the "List Profiles Extension" will be the same?
Will the Teleport implement the similar feature to openssh agent-restriction (https://www.openssh.com/agent-restrict.html#:~:text=Agent%20restriction%20in%20OpenSSH) ?

Reply:
I've updated this section in the RFD, but I'll answer each question here too.

> Will it be possible to list all end user's profiles, or will be there any restriction on forwarded certificates? 

Only the current profile at time of starting the remote session will be available through the forwarded agent.

> What happens in case of several hops, on every hop the output of the "List Profiles Extension" will be the same?

A user can continuously hop with the forwarded profile and certificates with `tsh ssh -A`, providing the same `key@goteleport.com` extension output on each hop.

> Will the Teleport implement the similar feature to openssh agent-restriction (https://www.openssh.com/agent-restrict.html#:~:text=Agent%20restriction%20in%20OpenSSH)?

This is a great idea, and it could potentially be used to safely forward a user's entire `~/.tsh` with enough restrictions in place. It's not currently planned, but I added a note in an additional consideration section for it.